File: tests/aes_test.py
from specs.aes import *
from sys import exit
import json

# mypy only checks functions that have types. So add an argument :)
def main(x: int) -> None:
    file = open('tests/test_vectors/aes128_test_vectors.json')
    aes128_test_vectors = json.load(file)
    for i in range(len(aes128_test_vectors)):
        msg = bytes.from_hex(aes128_test_vectors[i]['input'])
        k   = bytes.from_hex(aes128_test_vectors[i]['key'])
        n   = bytes.from_hex(aes128_test_vectors[i]['nonce'])
        ctr = int(aes128_test_vectors[i]['counter'],16)
        expected = bytes.from_hex(aes128_test_vectors[i]['output'])
        computed = aes128_encrypt_bytes(k,n,ctr,msg)
        if (computed == expected):
            print("Aes128 Test ",i," passed.")
        else:
            print("Aes128 Test ",i," failed:")
            print("expected ciphertext:",expected)
            print("computed ciphertext:",computed)
            exit(1)
    # rng = open("/dev/urandom","rb")
    # msgs = bytes.from_hex(rng.read(16 * 1024).hex())
    # key = bytes.from_hex(rng.read(16).hex())
    # nonce = bytes.from_hex(rng.read(12).hex())
    # ctr = uint32(1)
    # res = uint8(0)
    # for i in range(16):
    #     msg = msgs[i*1024:i*1024+1024]
    #     computed = aes128_encrypt(key,nonce,ctr,msg)
    #     res ^= computed[0]
    # print(res)


if __name__ == "__main__":
    main(0)

File: compiler/fstar-compiler/specs/tests/aes_test.ml
open Yojson.Basic.Util
open Testutil
   
let main () =
  let json = Yojson.Basic.from_file "../../../tests/test_vectors/aes128_test_vectors.json" in
  let jsonl = Yojson.Basic.Util.to_list json in
  List.iteri (fun i j ->
      let key = read_bytes j "key" in
      let nonce = read_bytes j  "nonce" in
      let counter = read_uint32 j "counter" in
      let input = read_bytes j "input" in
      let output = read_bytes j "output" in
      let computed = Aes.aes128_encrypt_bytes key nonce counter input in
      if output = computed then Printf.printf "Aes128 Test %d passed.\n" i
      else (Printf.printf "Aes128 Test %d failed.\n" i;
            Printf.printf "expected ciphertext: %s\n" (hex_of_bytes output);
            Printf.printf "computed ciphertext: %s\n" (hex_of_bytes computed));
      ()) jsonl;
  ()

let _ = main()
